primitives/core/src/bandersnatch.rs

```diff
@@ -0,0 +1,1016 @@
+// in the backend.
+const RING_CONTEXT_SERIALIZED_LEN: usize = 147752;
+
+/// Bandersnatch public key.
+#[cfg_attr(feature = "full_crypto", derive(Hash))]
+#[derive(
+	Clone,
+	Copy,
+	PartialEq,
+	Eq,
+	PartialOrd,
+	Ord,
+	Encode,
+	Decode,
+	PassByInner,
+	MaxEncodedLen,
+	TypeInfo,
+)]
+pub struct Public(pub [u8; PUBLIC_SERIALIZED_LEN]);
+
+impl UncheckedFrom<[u8; PUBLIC_SERIALIZED_LEN]> for Public {
+	fn unchecked_from(raw: [u8; PUBLIC_SERIALIZED_LEN]) -> Self {
+		Public(raw)
+	}
+}
+
+impl AsRef<[u8; PUBLIC_SERIALIZED_LEN]> for Public {
+	fn as_ref(&self) -> &[u8; PUBLIC_SERIALIZED_LEN] {
+		&self.0
+	}
+}
+
+impl AsRef<[u8]> for Public {
+	fn as_ref(&self) -> &[u8] {
+		&self.0[..]
+	}
+}
+
+impl AsMut<[u8]> for Public {
+	fn as_mut(&mut self) -> &mut [u8] {
+		&mut self.0[..]
+	}
+}
+
+impl TryFrom<&[u8]> for Public {
+	type Error = ();
+
+	fn try_from(data: &[u8]) -> Result<Self, Self::Error> {
+		if data.len() != PUBLIC_SERIALIZED_LEN {
+			return Err(())
+		}
+		let mut r = [0u8; PUBLIC_SERIALIZED_LEN];
+		r.copy_from_slice(data);
+		Ok(Self::unchecked_from(r))
+	}
+}
+
+impl ByteArray for Public {
+	const LEN: usize = PUBLIC_SERIALIZED_LEN;
+}
+
+impl TraitPublic for Public {}
+
+impl CryptoType for Public {
+	#[cfg(feature = "full_crypto")]
+	type Pair = Pair;
+}
+
+impl Derive for Public {}
+
+impl sp_std::fmt::Debug for Public {
+	#[cfg(feature = "std")]
+	fn fmt(&self, f: &mut sp_std::fmt::Formatter) -> sp_std::fmt::Result {
+		let s = self.to_ss58check();
+		write!(f, "{} ({}...)", crate::hexdisplay::HexDisplay::from(&self.as_ref()), &s[0..8])
+	}
+
+	#[cfg(not(feature = "std"))]
+	fn fmt(&self, _: &mut sp_std::fmt::Formatter) -> sp_std::fmt::Result {
+		Ok(())
+	}
+}
+
+/// Bandersnatch signature.
+///
+/// The signature is created via the [`VrfSecret::vrf_sign`] using [`SIGNING_CTX`] as `label`.
+#[cfg_attr(feature = "full_crypto", derive(Hash))]
+#[derive(Clone, Copy, PartialEq, Eq, Encode, Decode, PassByInner, MaxEncodedLen, TypeInfo)]
+pub struct Signature([u8; SIGNATURE_SERIALIZED_LEN]);
+
+impl UncheckedFrom<[u8; SIGNATURE_SERIALIZED_LEN]> for Signature {
+	fn unchecked_from(raw: [u8; SIGNATURE_SERIALIZED_LEN]) -> Self {
+		Signature(raw)
+	}
+}
+
+impl AsRef<[u8]> for Signature {
+	fn as_ref(&self) -> &[u8] {
+		&self.0[..]
+	}
+}
+
+impl AsMut<[u8]> for Signature {
+	fn as_mut(&mut self) -> &mut [u8] {
+		&mut self.0[..]
+	}
+}
+
+impl TryFrom<&[u8]> for Signature {
+	type Error = ();
+
+	fn try_from(data: &[u8]) -> Result<Self, Self::Error> {
+		if data.len() != SIGNATURE_SERIALIZED_LEN {
+			return Err(())
+		}
+		let mut r = [0u8; SIGNATURE_SERIALIZED_LEN];
+		r.copy_from_slice(data);
+		Ok(Self::unchecked_from(r))
+	}
+}
+
+impl ByteArray for Signature {
+	const LEN: usize = SIGNATURE_SERIALIZED_LEN;
+}
+
+impl CryptoType for Signature {
+	#[cfg(feature = "full_crypto")]
+	type Pair = Pair;
+}
+
+impl sp_std::fmt::Debug for Signature {
+	#[cfg(feature = "std")]
```

Comment:
Is this feature gate necessary? This should also work in no_std, no?

Reply:
I'll eventually modify this for the others primitives as well. Follow up PR